### ISSUE TITLE
[RFC] Bolt 3.3+ Extension Development Process

### ISSUE THREAD

**Issue by cdowdy** [bolt/bolt#6955]
I like to think I'm a fairly season developer :) but I ran into some "rough" patches migrating an extension from 3.2 to 3.3 following this guide https://docs.bolt.cm/3.3/extensions/basics/creating#creating-a-bolt-extension-or-theme  

Since I ran into them and I've been doing bolt extensions since v2 I figured others new to bolt would/could run into them too.  

1) Extensions used to be "local" for development. So a user could see this post https://docs.bolt.cm/3.3/upgrading/migrating-local-extensions-33 and think ok I have to move to a bundled extension to develop my extension for the marketplace. Apparently this isn't the case. and you should follow the link below in 2.   

2) Extensions for the marketplace need to follow https://docs.bolt.cm/3.3/extensions/basics/creating#creating-a-bolt-extension-or-theme    

For extensions already in the marketplace this I assume means I have to   

1) update to bolt 3.3.0+  
2) clone the extensions repo "*somewhere*" or  "*anywhere*"  
3) add the repo to  `/extensions/composer.json`  
  
```json   
 "repositories": {
        "packagist": false,
        "bolt": {
            "type": "composer",
            "url": "https://market.bolt.cm/satis/"
        },
        "myextension-git-repo": {
            "type": "path",
            "url": "/some/directory/anywhere/really/myextension/",
            "options": {
                "symlink": false
            }
        }
    },  
```  

Then update your minimum stability to dev and add it to the require section and run composer update. 

This will install your extension to `extensions/vendor/namespace/extension`.  

Now we have two directories. One our github repo we "work" in and another one that is created by composer.   

```   
- some directory or desktop  
  |--- git cloned extension repo  

├── bolt  /
│   ├── extensions /
│     ├── vendor  /
│       ├── namespace  /
│         └──your extension
```

How the extension is updated from our working github repo to the new composer created directory is completely opaque.  Do we:  

* manually copy all our work from the github repo to the vendor repo?
* manually copy all our work from the composer created extension to our github repo?   
* symlink these to each other?   

The issue is compounded by the composer update command. If you have a previous install with composer and update to 3.3 this creates a composer.lock file. You dump the autoloader in `extensions` and it's minimum stability is set back to "stable" instead of dev. Which is easy enough to fix be removing it.   

Whats the preferred workflow? How do we setup our extensions "right",  still be able to use git, not have multiple directories - or is this a necessary evil? - and see changes made to our extensions that we'll publish like we "used to" by simply refreshing the browser?   

Since I've just gotten started on 3.3+ this whole process seems like an awful lot to do when originally we had `extensions/local/namespace/extension` and could use git right there and load the extensions without having to jump through any extra steps besides manually copying our assets to the right directory.  I had to change a back end route in one place today for multiple extensions. (extend to extension) and this was almost enough for me to say "forget it" and lock the extensions below 3.3 haha.

**Comment by GawainLynch** [comment]
> You dump the autoloader in extensions and it's minimum stability is set back to "stable" instead of dev. Which is easy enough to fix be removing it.

For the record before this one gets going, this is actually controlled by the Composer settings in your `app/config/config.yml` file, and `extensions/composer.json` is updated either when visiting the Extensions UI page, or running Nut with one of the extension specific options.

**Comment by panda-madness** [comment]
The way I’ve setup extension development is like so (this is for 3.3+):

I have a bolt installation, called bolt-extensions-dev. In it’s extensions folder I create a subfolder for the extension that I want to make, init a git repo, and create a valid composer.json, according to the docs. I also load that extension through 3.3's bundled extensions system (meaning I add it as an autoload to the root composer.json, and register in .bolt.yml).

This way the extension is technically bundled (so no `composer update` shenanigans, you change the files and test immediatly), but can be published to the marketplace, provided you create a proper composer.json (you only have to manually test this once).

**Comment by cdowdy** [comment]
> For the record before this one gets going, this is actually controlled by the Composer settings in your app/config/config.yml file, 

I realized this after I went through the extension creation docs and got frustrated it kept going back to stable lol
  
  

> also load that extension through 3.3's bundled extensions system (meaning I add it as an autoload to the root composer.json, and register in .bolt.yml).

This works as long as your extension doesn't require any extra libraries. If you include others you then have to do all the classes, all the registering for each library and class you use. This could be big or tedious haha

**Comment by lenvanessen** [comment]
#6961

**Comment by cdowdy** [comment]
After taking a few days using the new extension development process and thanks to  @GawainLynch , @rossriley ,  and @SahAssar here's some other notes - please if you see something wrong don't hesitate to correct me I'm kind doing this by "trying" and theres bound to be something wrong or that I'm just not doing something obvious. 

I can answer these:  
  
> How the extension is updated from our working github repo to the new composer created directory is completely opaque. Do we:
> * manually copy all our work from the github repo to the vendor repo?
>*  manually copy all our work from the composer created extension to our github repo?
> * symlink these to each other?

None of these if you follow the "[official way](https://docs.bolt.cm/3.3/extensions/basics/creating#creating-a-bolt-extension-or-theme)" mentioned in the docs. You update your development extension by. 

1) Develop or fix you extension in your cloned / git initialized directory
2) push to git 
3) open a terminal, go to the extensions directory of you local Bolt install (not directory in 1 above )
4) run composer update in this directory  
5) repeat for each and __every__ change <sup>[1](#bolt-docs)</sup>.   
 
<b id="bolt-docs">1</b>  this isn't made clear in the docs.  

> From: https://docs.bolt.cm/3.3/extensions/basics/creating#installing-the-new-extension  
Once this is done, you can now either run composer update inside the extensions/ directory of your local Bolt site, or by pressing the "Run all Updates" button on the Extensions page in the Bolt back-end.  


> still be able to use git   

You can still use git you'll just more than likely have to change how you commit things to your repo or have a new branch setup specifically for development that you can then merge into a dev or master branch with a "cleaner" history. Since you have to commit everything (even something as simple as a missing period or typo you're going to have a slightly "messy" commit history. So finding what you exactly fixed or added or removed could be a challenge. This is something that I could probably improve on in my own workflow but I included it anyways :)  

> not have multiple directories - or is this a necessary evil?   

Nope. We have to have multiple directories. One to develop in, One that is created by composer. No longer are you able to develop test and push your extension in a local bolt install in one directory  

>  and see changes made to our extensions that we'll publish like we "used to" by simply refreshing the browser?    

You can still kinda but you'll need to run composer update after __each change__ you'll make and wait for it to hit git and for composer to recognize it, download it, remove your old one and install the changes.   

------  

Other quirks/rough edges:  
 
If you follow the docs and initialize or clone your extension repo in ``/home/user/me/extension`` and have a local install of bolt in ``/var/www/sites/bolt`` and use any bolt provided classes/methods you'll lose your ide autocomplete since it's not part of your local extension install - unless you setup your ide to tell it bolt is a Vendor library. Extensions more complicated than a ``Hello World`` type stink to develop this way. No longer will your IDE autocomplete methods or classes unless it's somehow included in your IDE as a library.   

As mentioned above you have to push each change to Git then run composer update to see any change. Gawain said in slack since bolt takes so much on it's own hands this is not easy to solve. Anecdotally this not exactly instant all the time. I changed a Heading title once and had to run composer update 3 times just to see the word ``and`` changed to an ampersand ``&amp``.  

So Thinking of  small changes here is what happens:  

1) change typo 
2) push to GitHub
3) open terminal/switch to new terminal tab composer update in Bolt install
4) switch back to browser and test change
5) If that's not it repeat 1-4 again.  

Compare this to the previous workflow:  

1) change typo
2) test in browser
3) if that's not it repeat 1-2 or push to branch if it is.   

Add in any complexity to your changes or if you simply need to dump something that a user is experiencing you have to push that dump just to see why they are getting what they are getting. That is to see  ``{{ dump( thing2Dump) }}`` we need to:  

1) put ``{{ dump( thing2Dump) }}`` in your template
2) push that change
3) update composer
4) reload browser to see dump call
5) attempt to debug the dump or underlying cause
6) push that change
7) composer update
8) test that change
9) repeat 1- 8 again for each change if needed.

All that context switching just to see and test a dump call - unless of course there is another way I don't know about haha. 

This is tedious for something as simple as a dump call. If you're used to a bunch of small commits this also clouds the git history as mentioned above. 

You can bypass all this but this also adds extra complexity to developing extensions or just an annoyance - mainly the last two. You can either: 

1) Mount and bind directories
2) Add a task runner that copies things from your local directory to your local bolt directory on change
3) Symlink your directories
4) Manually copy from develop to local bolt install or from local bolt to develop
5) Use a diff tool 

I understand the change was made to help people who use "local extensions" in their installs - which would now be bundled extensions for them - but it seems that the experience for extension/theme developers was made worse (especially if you have an extension that is more than a simple one) when you have to change your entire workflow just to update something as simple as a back end page's mount point / route. If you're building a theme and then do a bunch of color changes or layout changes to see how it looks this too is a bunch of steps just to see if the change you made looks right or is what you wanted.  

The thought of having to update my 5 extensions any of these ways makes me not want too haha and I'm struggling to remember anything else I've done to "extend" something that required all these steps. For other cms's I've done extensions for it was: install cms locally -> create extension directory in required spot -> develop -> test -> commit. This though too is anecdotal and just because something else does it a way doesn't mean it's good/right for everything.  

If I'm doing something bonkers or you read this and go "huh why you doing that ???" haha let me know because I don't want to have this feeling of dread when I need to update my extensions or fix something

**Comment by cdowdy** [comment]
ok. big ole update Thanks to @GawainLynch   

The "official way" is still the same as above:  

1) Install your extension/initialize a repo into a directory:  
  
   ``` /home/user/development/myextension/```  or windows: ```C:\User\You\development\Extension\```  

2) Add Bolt as a library in your IDE for number one above  
3) Develop your extension and push each change to git  
4) run composer update in the extensions directory of your local bolt install   

    ```/var/www/bolt/extensions  -> composer update ```    

5) repeat steps 3-4 for each change you make to your development extension. 

changes made to ``` /home/user/development/myextension/``` will now be mirrored to ```/var/www/bolt/extensions/vendor/yourNamespace/myextension```  

______
  
If you'd like to skip over all the pushing to git for each change you can either:  

1) copy your git initialized repo in the "official way" mentioned above to ```var/www/bolt/extensions/vendor/namespace/myextension/``` and move the git root to the vendor folder   or  

2) install your extension like normal -> remove all the contents of the ``````var/www/bolt/extensions/vendor/namespace/myextension/``` then either intitialize a repo there or clone your git repo there.    

If you do developemnt this way make sure you dont' run composer updaet or the extension's page "run all updates" you'll wipe out your "dev" extension   

____  

If you think this can be closed now go ahead :)

**Comment by SahAssar** [comment]
I think the above description summarizes it well enough, and future questions are probably best asked in #general in slack :)

**Comment by michaelborn** [comment]
> future questions are probably best asked in #general in slack

@SahAssar The slack channel has been great for developer questions, and you folks have been very kind to answer them, but in my opinion, good docs will save everyone lots of time and frustration. So from here on, expect (and solicit?) random documentation PRs. :)

**Comment by SvanteRichter** [comment]
@mborn319 We welcome all docs PRs, and we tried to enforce them for new features/changes for a while. The problem we saw was that it slowed feature development to a crawl in some cases since usually devs write good code and bad (or no) docs. I haven't given up hope on better docs, but the way there does not seem simple.
